### PR TITLE
feat: add progress bar

### PR DESCRIPTION
**Pull request by GrayFlash**
Adding a progress bar denoting the progress of `meteor run recipe` command.

image

Have used `github.com/cheggaaa/pb/` for the purpose. Let me know if yo have seen some other alternative based on plans around meteor!!

### CODE REVIEW

**Comment by ravisuhag** [comment]
@GrayFlash Check this also and see which one is better. 
https://github.com/schollz/progressbar

**Comment by GrayFlash** [comment]
> @GrayFlash Check this also and see which one is better. https://github.com/schollz/progressbar

Hey, checked both. Both are good in terms of number of people using them, and https://github.com/cheggaaa/pb/ is good in terms of a minimal progress bar with not many modifications needed.

But I think we should use https://github.com/schollz/progressbar cause it is more customizable in terms of color, themes, etc.

wdyt?? @ravisuhag

**Comment by ravisuhag** [comment]
> > @GrayFlash Check this also and see which one is better. https://github.com/schollz/progressbar
> 
> Hey, checked both. Both are good in terms of number of people using them, and https://github.com/cheggaaa/pb/ is good in terms of a minimal progress bar with not many modifications needed.
> 
> But I think we should use https://github.com/schollz/progressbar cause it is more customizable in terms of color, themes, etc.
> 
> wdyt?? @ravisuhag

Sounds good.

**Comment by GrayFlash** [comment]
Screenshot from 2022-01-18 11-11-10

Hey @ravisuhag updated the proogress bar.

**Comment by GrayFlash** [comment]
#223 mentions that "Meteor can be used in deployment mode. Meteor run command can accept a flag for daemon mode which does not report this and only uses standard logging." 

So what further changes are required for this complete issue??

**Comment by ravisuhag** [comment]
> #223 mentions that "Meteor can be used in deployment mode. Meteor run command can accept a flag for daemon mode which does not report this and only uses standard logging."
> 
> So what further changes are required for this complete issue??

@GrayFlash Ideally we would have the `isCI` function in the salt term package and use that to disable the progress bar in CI.

**Comment by StewartJingga** [comment]
> @GrayFlash Ideally we would have the isCI function in the salt term package and use that to disable the progress bar in CI.
> For now, we can add a flag to do it the non-interactive way which just disables the progress bar.
> 
> @StewartJingga WDYT?


+1, we could have a specific flag to disable this feature, like e.g. `-hide-progress-bar` to be more explicit, where if it is not defined will default to progress bar to be shown.

@GrayFlash @ravisuhag

File: cmd/run.go [hunk 1]
```diff
@@ -73,6 +74,13 @@ func RunCmd(lg log.Logger, mt *metrics.StatsdMonitor, cfg config.Config) *cobra.
 			var failures = 0
 			report = append(report, []string{"Status", "Recipe", "Source", "Duration(ms)", "Records"})
 
+			bar := progressbar.NewOptions(len(recipes),
+				progressbar.OptionEnableColorCodes(true),
+				progressbar.OptionSetDescription("[cyan]running recipes [reset]"),
+				progressbar.OptionShowCount(),
+			)
+			//.Default(int64(len(recipes)))
```
Comment (StewartJingga) [inline cmd/run.go:82]: don't forget to remove this line

**Review by StewartJingga — CHANGES_REQUESTED** [review]

**Comment by ravisuhag** [comment]
> > @GrayFlash Ideally we would have the isCI function in the salt term package and use that to disable the progress bar in CI.
> > For now, we can add a flag to do it the non-interactive way which just disables the progress bar.
> > @StewartJingga WDYT?
> 
> +1, we could have a specific flag to disable this feature, like e.g. `-hide-progress-bar` to be more explicit, where if it is not defined will default to progress bar to be shown.
> 
> @GrayFlash @ravisuhag

Let's call flag as `deamon` shortversion as `-d` ?

**Comment by StewartJingga** [comment]
> Let's call flag as daemon shortversion as -d ?

And if the flag is there Meteor only outputs logs? @ravisuhag

**Comment by ravisuhag** [comment]
> > Let's call flag as daemon shortversion as -d ?
> 
> 
> 
> And if the flag is there Meteor only outputs logs? @ravisuhag 

Yes

File: cmd/run.go (hunk 1 above)
Comment (GrayFlash) [reply cmd/run.go]: oh sorry, forgot to save the file before commit.. 😅

**Comment by kushsharma** [comment]
Optimus does it a little [differently](https://github.com/odpf/optimus/blob/main/cmd/progressbar.go), it checks if the output stream is a `tty`, if yes, only then shows the progress bar. It also supports taking an env variable `OPTIMUS_PROGRESS_INDICATOR` to disable it on demand.

Also, where are you streaming this progress bar? Is it `stdout` or `stderr`?

**Comment by ravisuhag** [comment]
> Optimus does it a little [differently](https://github.com/odpf/optimus/blob/main/cmd/progressbar.go), it [checks](https://github.com/odpf/optimus/blob/main/cmd/commands.go#L207) if the output stream is a `tty`, if yes, only then shows the progress bar. It also supports taking an env variable `OPTIMUS_PROGRESS_INDICATOR` to disable it on demand.
> 
> Also, where are you streaming this progress bar? Is it `stdout` or `stderr`?

+1 Have added a tty check function is [salt](https://github.com/odpf/salt/blob/9aa38ffc5626572e756f1abba7f1485b6e3de1d1/term/term.go#L10). 

We should also abstract out the progress bar as a lib in salt to standardize it across projects.

**Comment by ravisuhag** [comment]
@GrayFlash We can merge this for now. Later we will abstract out the progress bar in salt which also checks the TTY presence and use that. For now, it should be good.

**Comment by kushsharma** [comment]
Did we fixed the output of progress bar to be sent to `stderr` instead of `stdout`? If we are piping the output to a different command, like `grep`, you wouldn't want to mix the normal application output with progress bar.

**Comment by ravisuhag** [comment]
@kushsharma I am working on a PR to abstract that out in salt and we will use that here. PR coming soon. So merged this for now.

**Pull request merged**